# 02_Spark_Dataframe

#### Ref:
1. Spark official link: https://spark.apache.org/
2. Spark Python API: https://spark.apache.org/docs/latest/api/python/reference/index.html

#### How to list all available datasets from databricks cluster

In [0]:
%fs ls /databricks-datasets

path name size modificationTime dbfs:/databricks-datasets/ databricks-datasets/ 0 0 dbfs:/databricks-datasets/COVID/ COVID/ 0 0 dbfs:/databricks-datasets/README.md README.md 976 1532468253000 dbfs:/databricks-datasets/Rdatasets/ Rdatasets/ 0 0 dbfs:/databricks-datasets/SPARK_README.md SPARK_README.md 3359 1455043490000 dbfs:/databricks-datasets/adult/ adult/ 0 0 dbfs:/databricks-datasets/airlines/ airlines/ 0 0 dbfs:/databricks-datasets/amazon/ amazon/ 0 0 dbfs:/databricks-datasets/asa/ asa/ 0 0 dbfs:/databricks-datasets/atlas_higgs/ atlas_higgs/ 0 0 dbfs:/databricks-datasets/bikeSharing/ bikeSharing/ 0 0 dbfs:/databricks-datasets/cctvVideos/ cctvVideos/ 0 0 dbfs:/databricks-datasets/credit-card-fraud/ credit-card-fraud/ 0 0 dbfs:/databricks-datasets/cs100/ cs100/ 0 0 dbfs:/databricks-datasets/cs110x/ cs110x/ 0 0 dbfs:/databricks-datasets/cs190/ cs190/ 0 0 dbfs:/databricks-datasets/data.gov/ data.gov/ 0 0 dbfs:/databricks-datasets/definitive-guide/ definitive-guide/ 0 0 dbfs:/databricks-datasets/delta-sharing/ delta-sharing/ 0 0 dbfs:/databricks-datasets/flights/ flights/ 0 0 dbfs:/databricks-datasets/flower_photos/ flower_photos/ 0 0 dbfs:/databricks-datasets/flowers/ flowers/ 0 0 dbfs:/databricks-datasets/genomics/ genomics/ 0 0 dbfs:/databricks-datasets/hail/ hail/ 0 0 dbfs:/databricks-datasets/identifying-campaign-effectiveness/ identifying-campaign-effectiveness/ 0 0 dbfs:/databricks-datasets/iot/ iot/ 0 0 dbfs:/databricks-datasets/iot-stream/ iot-stream/ 0 0 dbfs:/databricks-datasets/learning-spark/ learning-spark/ 0 0 dbfs:/databricks-datasets/learning-spark-v2/ learning-spark-v2/ 0 0 dbfs:/databricks-datasets/lending-club-loan-stats/ lending-club-loan-stats/ 0 0 dbfs:/databricks-datasets/med-images/ med-images/ 0 0 dbfs:/databricks-datasets/media/ media/ 0 0 dbfs:/databricks-datasets/mnist-digits/ mnist-digits/ 0 0 dbfs:/databricks-datasets/news20.binary/ news20.binary/ 0 0 dbfs:/databricks-datasets/nyctaxi/ nyctaxi/ 0 0 dbfs:/databricks-datasets/nyctaxi-with-zipcodes/ nyctaxi-with-zipcodes/ 0 0 dbfs:/databricks-datasets/online_retail/ online_retail/ 0 0 dbfs:/databricks-datasets/overlap-join/ overlap-join/ 0 0 dbfs:/databricks-datasets/power-plant/ power-plant/ 0 0 dbfs:/databricks-datasets/retail-org/ retail-org/ 0 0 dbfs:/databricks-datasets/rwe/ rwe/ 0 0 dbfs:/databricks-datasets/sai-summit-2019-sf/ sai-summit-2019-sf/ 0 0 dbfs:/databricks-datasets/sample_logs/ sample_logs/ 0 0 dbfs:/databricks-datasets/samples/ samples/ 0 0 dbfs:/databricks-datasets/sfo_customer_survey/ sfo_customer_survey/ 0 0 dbfs:/databricks-datasets/sms_spam_collection/ sms_spam_collection/ 0 0 dbfs:/databricks-datasets/songs/ songs/ 0 0 dbfs:/databricks-datasets/structured-streaming/ structured-streaming/ 0 0 dbfs:/databricks-datasets/timeseries/ timeseries/ 0 0 dbfs:/databricks-datasets/tpch/ tpch/ 0 0 dbfs:/databricks-datasets/warmup/ warmup/ 0 0 dbfs:/databricks-datasets/weather/ weather/ 0 0 dbfs:/databricks-datasets/wiki/ wiki/ 0 0 dbfs:/databricks-datasets/wikipedia-datasets/ wikipedia-datasets/ 0 0 dbfs:/databricks-datasets/wine-quality/ wine-quality/ 0 0

**spark is a spark session object**, and every spark program starts with spark session object.

In [0]:
spark

SparkSession - hive 
 
 
 SparkContext 

 Spark UI 

 
 Version 
 v3.2.1 
 Master 
 local[8] 
 AppName 
 Databricks Shell

#### Spark Dataframe 
2. **read** is an attribute of spark session object and it returns dataframe using format, option, load
3. **format**: mentioning to dataframe reader that dataframe is a csv file.
4. **option**: also to reader, csv file is comming with header row and it should follow inferShema
5. **load**: reader will read the data from the location.

In [0]:
fire_df = spark.read \
            .format("csv") \
            .option("header", "true") \
            .option("inferSchema", "true") \
            .load("/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv")


##### csv is a shortcut method, but prefferd to use above long form to read any type of file as its standardized

In [0]:
fire_df = spark.read \
            .csv("/databricks-datasets/learning-spark-v2/sf-fire/sf-fire-calls.csv", 
                header=True,
                inferSchema=True)

##### show() is not retuning dataframe nice way,

In [0]:
fire_df.show(10)

+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+--------------------+----+-------------------+---------+------------+----+------------+--------+--------------+--------+---------------+---------+------------+------------------------------+------------------------+-------------------+---------------+--------------------+-------------+------------------+
|Call Number|Unit ID|Incident Number|        CallType| Call Date|Watch Date|Call Final Disposition|      Available DtTm|             Address|City|Zipcode of Incident|Battalion|Station Area| Box|OrigPriority|Priority|Final Priority|ALS Unit|Call Type Group|NumAlarms|    UnitType|Unit sequence in call dispatch|Fire Prevention District|Supervisor District|   Neighborhood|            Location|        RowID|             Delay|
+-----------+-------+---------------+----------------+----------+----------+----------------------+--------------------+--------------------+----+--

##### display() is showing dataframe in a tabular format. 
##### but its not belongs to a apache spark. Its a optional tool of databricks.

In [0]:
display(fire_df.limit(10))

Call Number Unit ID Incident Number CallType Call Date Watch Date Call Final Disposition Available DtTm Address City Zipcode of Incident Battalion Station Area Box OrigPriority Priority Final Priority ALS Unit Call Type Group NumAlarms UnitType Unit sequence in call dispatch Fire Prevention District Supervisor District Neighborhood Location RowID Delay 20110014 M29 2003234 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 01:58:43 AM 10TH ST/MARKET ST SF 94103 B02 36 2338 1 1 2 true null 1 MEDIC 1 2 6 Tenderloin (37.7765408927183, -122.417501464907) 020110014-M29 5.233333333333333 20110015 M08 2003233 Medical Incident 01/11/2002 01/10/2002 Other 01/11/2002 02:10:17 AM 300 Block of 5TH ST SF 94107 B03 08 2243 1 1 2 true null 1 MEDIC 1 3 6 South of Market (37.7792841462441, -122.402061300134) 020110015-M08 3.0833333333333335 20110016 B02 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 6 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B02 3.05 20110016 B04 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:54 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 3 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-B04 2.316666666666667 20110016 D2 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 CHIEF 4 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-D2 3.0166666666666666 20110016 E03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.683333333333333 20110016 E38 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:51:17 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 1 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E38 2.1 20110016 E41 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:47:00 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 8 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E41 2.716666666666667 20110016 M03 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:38 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 true null 1 MEDIC 10 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-M03 2.7666666666666666 20110016 RS1 2003235 Structure Fire 01/11/2002 01/10/2002 Other 01/11/2002 01:46:57 AM 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 RESCUE SQUAD 9 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-RS1 3.2666666666666666

In [0]:
fire_df.columns

Out[6]: ['Call Number',
 'Unit ID',
 'Incident Number',
 'CallType',
 'Call Date',
 'Watch Date',
 'Call Final Disposition',
 'Available DtTm',
 'Address',
 'City',
 'Zipcode of Incident',
 'Battalion',
 'Station Area',
 'Box',
 'OrigPriority',
 'Priority',
 'Final Priority',
 'ALS Unit',
 'Call Type Group',
 'NumAlarms',
 'UnitType',
 'Unit sequence in call dispatch',
 'Fire Prevention District',
 'Supervisor District',
 'Neighborhood',
 'Location',
 'RowID',
 'Delay']

In [0]:
fire_df.dtypes

Out[11]: [('Call Number', 'int'),
 ('Unit ID', 'string'),
 ('Incident Number', 'int'),
 ('CallType', 'string'),
 ('Call Date', 'string'),
 ('Watch Date', 'string'),
 ('Call Final Disposition', 'string'),
 ('Available DtTm', 'string'),
 ('Address', 'string'),
 ('City', 'string'),
 ('Zipcode of Incident', 'int'),
 ('Battalion', 'string'),
 ('Station Area', 'string'),
 ('Box', 'string'),
 ('OrigPriority', 'string'),
 ('Priority', 'string'),
 ('Final Priority', 'int'),
 ('ALS Unit', 'boolean'),
 ('Call Type Group', 'string'),
 ('NumAlarms', 'int'),
 ('UnitType', 'string'),
 ('Unit sequence in call dispatch', 'int'),
 ('Fire Prevention District', 'string'),
 ('Supervisor District', 'string'),
 ('Neighborhood', 'string'),
 ('Location', 'string'),
 ('RowID', 'string'),
 ('Delay', 'double')]

##### createGlobalTempView
1. Creating global temporary view similar like a table.
2. we can query through sql in this view

In [0]:
fire_df.createGlobalTempView("fire_service_calls_view")

##### global_temp is a hidden database
1. all global temporary tables, views are created inside it.

In [0]:
%sql

select distinct `Call Date` from global_temp.fire_service_calls_view limit 10

Call Date 01/17/2002 01/14/2002 01/15/2002 01/20/2002 01/12/2002 01/16/2002 01/18/2002 01/11/2002 01/13/2002 01/19/2002

## Dataframe Transformations

##### Rename Spark Dataframe Column

In [0]:
list_columns = fire_df.columns
for col in list_columns:
    print(f'.withColumnRenamed("{col}", "{col.replace(" ","")}")  \\')       

.withColumnRenamed("Call Number", "CallNumber")  \
.withColumnRenamed("Unit ID", "UnitID")  \
.withColumnRenamed("Incident Number", "IncidentNumber")  \
.withColumnRenamed("CallType", "CallType")  \
.withColumnRenamed("Call Date", "CallDate")  \
.withColumnRenamed("Watch Date", "WatchDate")  \
.withColumnRenamed("Call Final Disposition", "CallFinalDisposition")  \
.withColumnRenamed("Available DtTm", "AvailableDtTm")  \
.withColumnRenamed("Address", "Address")  \
.withColumnRenamed("City", "City")  \
.withColumnRenamed("Zipcode of Incident", "ZipcodeofIncident")  \
.withColumnRenamed("Battalion", "Battalion")  \
.withColumnRenamed("Station Area", "StationArea")  \
.withColumnRenamed("Box", "Box")  \
.withColumnRenamed("OrigPriority", "OrigPriority")  \
.withColumnRenamed("Priority", "Priority")  \
.withColumnRenamed("Final Priority", "FinalPriority")  \
.withColumnRenamed("ALS Unit", "ALSUnit")  \
.withColumnRenamed("Call Type Group", "CallTypeGroup")  \
.withColumnRenamed("NumAlarms",

In [0]:
renamed_fire_df = fire_df \
                    .withColumnRenamed("Call Number", "CallNumber")  \
                    .withColumnRenamed("Unit ID", "UnitID")  \
                    .withColumnRenamed("Incident Number", "IncidentNumber")  \
                    .withColumnRenamed("CallType", "CallType")  \
                    .withColumnRenamed("Call Date", "CallDate")  \
                    .withColumnRenamed("Watch Date", "WatchDate")  \
                    .withColumnRenamed("Call Final Disposition", "CallFinalDisposition")  \
                    .withColumnRenamed("Available DtTm", "AvailableDtTm")  \
                    .withColumnRenamed("Address", "Address")  \
                    .withColumnRenamed("City", "City")  \
                    .withColumnRenamed("Zipcode of Incident", "ZipcodeofIncident")  \
                    .withColumnRenamed("Battalion", "Battalion")  \
                    .withColumnRenamed("Station Area", "StationArea")  \
                    .withColumnRenamed("Box", "Box")  \
                    .withColumnRenamed("OrigPriority", "OrigPriority")  \
                    .withColumnRenamed("Priority", "Priority")  \
                    .withColumnRenamed("Final Priority", "FinalPriority")  \
                    .withColumnRenamed("ALS Unit", "ALSUnit")  \
                    .withColumnRenamed("Call Type Group", "CallTypeGroup")  \
                    .withColumnRenamed("NumAlarms", "NumAlarms")  \
                    .withColumnRenamed("UnitType", "UnitType")  \
                    .withColumnRenamed("Unit sequence in call dispatch", "Unitsequenceincalldispatch")  \
                    .withColumnRenamed("Fire Prevention District", "FirePreventionDistrict")  \
                    .withColumnRenamed("Supervisor District", "SupervisorDistrict")  \
                    .withColumnRenamed("Neighborhood", "Neighborhood")  \
                    .withColumnRenamed("Location", "Location")  \
                    .withColumnRenamed("RowID", "RowID")  \
                    .withColumnRenamed("Delay", "Delay")

##### Check the Spark Dataframe columns and its datatype

In [0]:
renamed_fire_df.dtypes

Out[34]: [('CallNumber', 'int'),
 ('UnitID', 'string'),
 ('IncidentNumber', 'int'),
 ('CallType', 'string'),
 ('CallDate', 'string'),
 ('WatchDate', 'string'),
 ('CallFinalDisposition', 'string'),
 ('AvailableDtTm', 'string'),
 ('Address', 'string'),
 ('City', 'string'),
 ('ZipcodeofIncident', 'int'),
 ('Battalion', 'string'),
 ('StationArea', 'string'),
 ('Box', 'string'),
 ('OrigPriority', 'string'),
 ('Priority', 'string'),
 ('FinalPriority', 'int'),
 ('ALSUnit', 'boolean'),
 ('CallTypeGroup', 'string'),
 ('NumAlarms', 'int'),
 ('UnitType', 'string'),
 ('Unitsequenceincalldispatch', 'int'),
 ('FirePreventionDistrict', 'string'),
 ('SupervisorDistrict', 'string'),
 ('Neighborhood', 'string'),
 ('Location', 'string'),
 ('RowID', 'string'),
 ('Delay', 'double')]

In [0]:
renamed_fire_df.printSchema()

root
 |-- CallNumber: integer (nullable = true)
 |-- UnitID: string (nullable = true)
 |-- IncidentNumber: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- CallDate: string (nullable = true)
 |-- WatchDate: string (nullable = true)
 |-- CallFinalDisposition: string (nullable = true)
 |-- AvailableDtTm: string (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- ZipcodeofIncident: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- StationArea: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- FinalPriority: integer (nullable = true)
 |-- ALSUnit: boolean (nullable = true)
 |-- CallTypeGroup: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unitsequenceincalldispatch: integer (nullable = true)
 |-- FirePreventionDistrict: string (nullable = t

In [0]:
from pyspark.sql.functions import to_date, to_timestamp, round

fire_df = renamed_fire_df \
            .withColumn("CallDate", to_date("CallDate", "MM/dd/yyyy"))  \
            .withColumn("WatchDate", to_date("WatchDate", "MM/dd/yyyy"))  \
            .withColumn("AvailableDtTm", to_timestamp("AvailableDtTm", "MM/dd/yyyy hh:mm:ss a"))  \
            .withColumn("Delay", round("Delay", 2))

In [0]:
display(fire_df.sample(0.1).limit(10))

CallNumber UnitID IncidentNumber CallType CallDate WatchDate CallFinalDisposition AvailableDtTm Address City ZipcodeofIncident Battalion StationArea Box OrigPriority Priority FinalPriority ALSUnit CallTypeGroup NumAlarms UnitType Unitsequenceincalldispatch FirePreventionDistrict SupervisorDistrict Neighborhood Location RowID Delay 20110016 E03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 ENGINE 7 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-E03 2.68 20110016 T03 2003235 Structure Fire 2002-01-11 2002-01-10 Other 2002-01-11T01:47:00.000+0000 2000 Block of CALIFORNIA ST SF 94109 B04 38 3362 3 3 3 false null 1 TRUCK 5 4 5 Pacific Heights (37.7895840679362, -122.428071912459) 020110016-T03 3.23 20110021 E03 2003240 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T02:17:23.000+0000 300 Block of ELLIS ST SF 94102 B03 03 1461 3 3 3 false null 1 ENGINE 1 3 6 Tenderloin (37.7850246606887, -122.411971890566) 020110021-E03 2.93 20110028 E21 2003247 Vehicle Fire 2002-01-11 2002-01-10 Other 2002-01-11T03:55:07.000+0000 1000 Block of OAK ST SF 94117 B05 21 4141 3 3 3 false null 1 ENGINE 1 5 5 Hayes Valley (37.7732938420762, -122.436598149618) 020110028-E21 1.88 20110029 M08 2003248 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T04:02:46.000+0000 800 Block of BRYANT ST SF 94103 B03 08 2255 3 3 3 true null 1 MEDIC 1 3 6 South of Market (37.775420706711, -122.403404791479) 020110029-M08 2.45 20110035 M32 2003253 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T04:59:27.000+0000 1200 Block of CORTLAND AVE SF 94110 B10 32 5725 3 3 3 true null 1 MEDIC 2 10 9 Bernal Heights (37.7397966983548, -122.411715561027) 020110035-M32 4.48 20110038 E12 2003256 Alarms 2002-01-11 2002-01-10 Other 2002-01-11T05:37:27.000+0000 400 Block of PARNASSUS AVE SF 94131 B05 12 5155 3 3 3 false null 1 ENGINE 2 5 7 Inner Sunset (37.7633165969878, -122.458205988587) 020110038-E12 3.83 20110060 E36 2003270 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T07:24:53.000+0000 100 Block of VAN NESS AVE SF 94102 B02 36 3166 3 3 3 false null 1 ENGINE 1 2 6 Tenderloin (37.7768524577717, -122.419455012102) 020110060-E36 3.95 20110063 T10 2003273 Alarms 2002-01-11 2002-01-10 Other 2002-01-11T07:25:06.000+0000 100 Block of JORDAN AVE SF 94118 B07 10 4452 3 3 3 false null 1 TRUCK 3 7 2 Presidio Heights (37.7826808102254, -122.456790301039) 020110063-T10 2.9 20110066 93 2003274 Medical Incident 2002-01-11 2002-01-10 Other 2002-01-11T08:21:08.000+0000 600 Block of VAN NESS AVE SF 94102 B02 36 3164 2 2 2 false null 1 MEDIC 1 2 5 Western Addition (37.78147394156, -122.42068958116) 020110066-93 9.1

In [0]:
fire_df.printSchema()

root
 |-- CallNumber: integer (nullable = true)
 |-- UnitID: string (nullable = true)
 |-- IncidentNumber: integer (nullable = true)
 |-- CallType: string (nullable = true)
 |-- CallDate: date (nullable = true)
 |-- WatchDate: date (nullable = true)
 |-- CallFinalDisposition: string (nullable = true)
 |-- AvailableDtTm: timestamp (nullable = true)
 |-- Address: string (nullable = true)
 |-- City: string (nullable = true)
 |-- ZipcodeofIncident: integer (nullable = true)
 |-- Battalion: string (nullable = true)
 |-- StationArea: string (nullable = true)
 |-- Box: string (nullable = true)
 |-- OrigPriority: string (nullable = true)
 |-- Priority: string (nullable = true)
 |-- FinalPriority: integer (nullable = true)
 |-- ALSUnit: boolean (nullable = true)
 |-- CallTypeGroup: string (nullable = true)
 |-- NumAlarms: integer (nullable = true)
 |-- UnitType: string (nullable = true)
 |-- Unitsequenceincalldispatch: integer (nullable = true)
 |-- FirePreventionDistrict: string (nullable = tr

## Data Analysis on Fire Calls Dataframe

#### Cache the Spark Dataframe

In [0]:
fire_df.cache()

Out[15]: DataFrame[CallNumber: int, UnitID: string, IncidentNumber: int, CallType: string, CallDate: date, WatchDate: date, CallFinalDisposition: string, AvailableDtTm: timestamp, Address: string, City: string, ZipcodeofIncident: int, Battalion: string, StationArea: string, Box: string, OrigPriority: string, Priority: string, FinalPriority: int, ALSUnit: boolean, CallTypeGroup: string, NumAlarms: int, UnitType: string, Unitsequenceincalldispatch: int, FirePreventionDistrict: string, SupervisorDistrict: string, Neighborhood: string, Location: string, RowID: string, Delay: double]

##### Q1. How many distinct types of calls were made to the Fire Department?
```sql
select count(distinct CallType) as distinct_call_type_count 
from demo_db.fire_service_calls_tbl
where CallType is not null
```

In [0]:
fire_df.createOrReplaceTempView("fire_service_calls_view")
q1_sql_df = spark.sql("""
            select count(distinct CallType) as distinct_call_type_count 
            from fire_service_calls_view
            where CallType is not null
        """)
display(q1_sql_df)

distinct_call_type_count 32

In [0]:
q1_df = fire_df.where("CallType is not null")  \
                .select("CallType")  \
                .distinct()
print(q1_df.count())

32


##### Q2. What are the distinct types of calls made to the fire department?
```sql
select distinct CallType as distinct_call_type 
from demo_db.fire_service_calls_tbl
where CallType is not null
```

In [0]:
from pyspark.sql.functions import expr
q2_df = fire_df.where("CallType is not null")  \
                .select(expr("CallType as distinct_CallType"))  \
                .distinct()
display(q2_df.limit(10))

distinct_CallType Alarms Odor (Strange / Unknown) Citizen Assist / Service Call Vehicle Fire Other Outside Fire Structure Fire Medical Incident Fuel Spill Smoke Investigation (Outside)

##### Q3. Find out all responses for delayed times greater than 5 mints?
```sql
select CallNumber, Delay
from demo_db.fire_service_calls_tbl
where Delay > 5
```

In [0]:
q3_df = fire_df.where("Delay > 5")  \
                .select(["CallNumber", "Delay"])
display(q3_df.limit(10))

CallNumber Delay 20110014 5.23 20110017 6.93 20110019 6.12 20110039 7.85 20110045 77.33 20110046 5.42 20110055 6.5 20110058 6.85 20110058 6.85 20110061 6.33

##### Q4. What were the most common call types?
```sql
select CallType, count(*) as count
from demo_db.fire_service_calls_tbl
where CallType is not null
group by CallType
order by count desc
```

1. count is action or transformation?
    1. DataFrame.count() -> Action
    2. GroupedData.count() -> Transformation

In [0]:
fire_df.select("CallType")  \
    .where("CallType is not null")  \
    .groupBy("CallType") \
    .count() \
    .orderBy("count", ascending=False) \
    .show(10)  # showing top 10 records only
    

+--------------------+-------+
|            CallType|  count|
+--------------------+-------+
|    Medical Incident|2843475|
|      Structure Fire| 578998|
|              Alarms| 483518|
|   Traffic Collision| 175507|
|Citizen Assist / ...|  65360|
|               Other|  56961|
|        Outside Fire|  51603|
|        Vehicle Fire|  20939|
|        Water Rescue|  20037|
|Gas Leak (Natural...|  17284|
+--------------------+-------+
only showing top 10 rows



##### Q5. What zip codes accounted for the most common calls?
```sql
select CallType, zipcode, count(*) as count
from demo_db.fire_service_calls_tbl
where CallType is not null
group by CallType, zipcode
order by count desc
```

##### Q6. What San Francisco neighborhoods are in zip codes 94102 and 94103
```sql
select zipcode, Neighborhood
from demo_db.fire_service_calls_tbl
where zipcode == 94102 or zipcode == 94103
```

##### Q7. What was the sum of all call alarms, average, min and max of the call response times?
```sql
select Sum(NumAlarms) as sum_call_alarms, 
avg(delay) as avg_delay, 
min(delay) as min_delay, 
max(delay) as max_delay
from demo_db.fire_service_calls_tbl
```

##### Q8. How many distinct years of data are in csv files?
```sql
select distinct year(to_date(CallDate, "MM/dd/yyyy")) as distinct_year
from demo_db.fire_service_calls_tbl
```

##### Q9. What week of the year in 2018 had the most fire calls?
```sql
select weekofyear(to_date(CallDate, "MM/dd/yyyy")) as week_year, count(*) as count 
from demo_db.fire_service_calls_tbl
where year(to_date(CallDate, "MM/dd/yyyy")) = 2018
group by week_year
order by count desc
```

##### Q10. What neighborhoods in San Francisco had the worst response time in 2018?
```sql
select Neighborhood, delay
from demo_db.fire_service_calls_tbl
where year(to_date(CallDate, "MM/dd/yyyy")) = 2018
order by delay desc
```